### DISCUSSION TITLE
How to open SpreadsheetDocument without calcchain.xml

### DISCUSSION THREAD

**Discussion by Distancehe** [OfficeDev/Open-XML-SDK#1279] [Q&A]
The Excel file itself do not contain calcchain.xml, but calcchain.xml is referenced from workbook.xml.rels as below snapshot. When trying to load the excel with SpreadsheetDocument.Open(), it will throw exception since the calcchain.xml is missing. The question is, how to open the excel file with calcchain.xml missing? Is there some way to ignore this and continue open the file ? Thanks !

![image](https://user-images.githubusercontent.com/4394577/209112903-02252d47-95a6-4be0-a404-b3ff1ca45c63.png)

**Comment by tomjebo** [comment]
@Distancehe Can you provide the sample document that is failing to open in Excel and also explain how this file was produced? Our implementer notes don't say anything about the requirement for the calcChain root or it's containing part calcchain.xml so I would have to research this a little to find out if it's supposed to be valid or not. But in the meantime, it would help to know how the workbook was created.

**Reply by Distancehe** [reply]
@tomjebo thanks for the response ! This is the sample file and it generated from service with POI https://poi.apache.org/

**Reply by Distancehe** [reply]
When I create an empty calcChain.xml and add to the excel file, the document can be opened successfully. But I think this more like a workaround, so still want for official solution.

Empty calcChain.xml content is just like below,
![image](https://user-images.githubusercontent.com/4394577/209303550-b0550455-a31f-4929-b289-f80d4c0309fa.png)

**Reply by Distancehe** [reply]
From https://learn.microsoft.com/en-us/office/open-xml/working-with-the-calculation-chain, "While calculation chain information can be loaded by a spreadsheet application, **it is not required.** A calculation chain can be constructed in memory at load-time based on the formulas and their interdependence, if the spreadsheet application finds this information useful."
Can we bypass the checking for calcChain.xml while SpreadsheetDocument.Open() the spreadsheet ?

**Comment by tomjebo** [comment]
@Distancehe so Excel will also add the calcChain.xml part back into the package on save, but with some kind of "correct" content. After looking at what Excel adds for the content, I'm not sure having SpreadsheetDocument.Open() add the missing part is the right way to go. Ignoring it might be possible but that would leave a missing part relationship and potentially important information that *should* have been added by the producer of the package. Excel is essentially saying that it can recreate the missing information but the SDK wouldn't have the logic that Excel has. (That's why it's Excel). Having said that, I will talk to our Excel team to find out what the ramifications of adding an empty calcChain.xml part (because the SDK doesn't want to process the whole workbook looking for formulas and adding cells) would be. For example, after opening with a blank calcChain.xml inserted but before saving the workbook with Excel, could it lead to incorrect formula computations?

**Reply by Distancehe** [reply]
Thanks for the reply ! Actually, I'm also seeking for approach to ignore calcChain.xml with SpreadsheetDocument.Open() but do not know how to achieve it, any suggestion would be appreciated !
Also, by adding the mock calcChain.xml and open with SDK, it will not load formula related computations, that mock file is just a workaround to make SpreadsheetDocument.Open() load the Spreadsheet.

**Reply by tomjebo** [reply]
So I did have a discussion with the Excel team on this and their take was that the SDK is correct to be flagging this as it technically represents a corrupt package. They said that originally, the calcChain part is not a requirement for 3rd party producers and Excel typically adds this so it is a bit of an odd case. Their recommendation was actually to remove the relationship vs adding a part to complete it. So in light of this I think the SDK could catch the missing part and remove that relationship upon Open().  I will look into this and submit a PR unless you are interesting in contributing.

**Reply by Distancehe** [reply]
To confirm, "Their recommendation was actually to remove the relationship vs adding a part to complete it.", is it mean to remove below relationship from workbook.xml.rels when Open() ? If so, by open the file with Excel app, will it add the calcChain.xml back if there's formula in the sheet ?

Relationship Id="rIdx" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/calcChain" Target="calcChain.xml"

**Reply by tomjebo** [reply]
Yes, the recommendation was to remove the relationship to correct the package, not necessarily in the Open() but that's where we might be able to do it. And they did say that Excel would add calcChain parts if they are needed.

**Reply by Distancehe** [reply]
I see. So even without the relationship definition from rel file, Excel can add the calcChain.xml back and make the formula work as expected ?

**Reply by tomjebo** [reply]
If the part was missing to begin with then there's not much choice anyway. But yes, the Excel team says that the recommendation to 3rd party authors has been not to write a calcChain part or relationship and let Excel take care of this. Excel is able to understand all dependencies as well as optimizations and concurrency/parallel options of subchains. The chain will actually change as partial and full calculations are done by Excel until it reaches a steady state which is optimal. You can see some description of this in ISO 29500-1:2016 L.2.5 Calculation Chain. From the overview of that section: _"In other words, the Calculation Chain part does not indicate the
dependencies that formulas have on other cell values; it only indicates the order in which the cells were
last calculated"_ Excel does this processing.

**Reply by Distancehe** [reply]
Got it. Thanks for the explanation and help all along !

**Comment by tomjebo** [comment]
@Distancehe I've submitted a fix in PR #1288, please test if you have time. This does leave the relationship intact vs removing it as I had mentioned before. I can rework if you think it should remove the relationship entry.

**Reply by Distancehe** [reply]
Thanks @tomjebo !
